app/feature/feature-insurances/src/test/kotlin/com/hedvig/android/feature/insurances/terminatedcontracts/TerminatedContractsPresenterTest.kt

```diff
@@ -0,0 +1,351 @@
+  val testLogcatLogger = TestLogcatLoggingRule()
+
+  @Test
+  fun `if there are no terminated insurances we get no terminated insurances state`() = runTest {
+    val getInsuranceContractsUseCase = FakeGetInsuranceContractsUseCase()
+    val presenter = TerminatedContractsPresenter { getInsuranceContractsUseCase }
+    presenter.test(TerminatedContractsUiState.Loading) {
+      assertThat(awaitItem()).isEqualTo(TerminatedContractsUiState.Loading)
+      getInsuranceContractsUseCase.addOnlyActiveInsurancesToResponse()
+      assertThat(awaitItem()).isEqualTo(
+        TerminatedContractsUiState.NoTerminatedInsurances,
+      )
+    }
+  }
+
+  @Test
+  fun `with an initial success state, if there is an error, we are able to retry and get back in success state again`() =
+    runTest {
+      val getInsuranceContractsUseCase = FakeGetInsuranceContractsUseCase()
+      val presenter = TerminatedContractsPresenter { getInsuranceContractsUseCase }
+      presenter.test(TerminatedContractsUiState.Success(getInsuranceContractsUseCase.getTerminatedInsurances())) {
+        assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Success>()
+        getInsuranceContractsUseCase.addErrorToResponse()
+        assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Error>()
+        sendEvent(TerminatedContractsEvent.Retry)
+        assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Loading>()
+        getInsuranceContractsUseCase.addTerminatedInsurancesToResponse()
+        assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Success>()
+      }
+    }
+
+  @Test
+  fun `with an initial error state, we are able to retry and if response is successful show success state`() = runTest {
+    val getInsuranceContractsUseCase = FakeGetInsuranceContractsUseCase()
+    val presenter = TerminatedContractsPresenter { getInsuranceContractsUseCase }
+    presenter.test(TerminatedContractsUiState.Error) {
+      assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Error>()
+      sendEvent(TerminatedContractsEvent.Retry)
+      assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Loading>()
+      getInsuranceContractsUseCase.addTerminatedInsurancesToResponse()
+      assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Success>()
+    }
+  }
+
+  @Test
+  fun `with an initial error state, if there comes a successful response in flow, show success state`() = runTest {
+    val getInsuranceContractsUseCase = FakeGetInsuranceContractsUseCase()
+    val presenter = TerminatedContractsPresenter { getInsuranceContractsUseCase }
+    presenter.test(TerminatedContractsUiState.Error) {
+      assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Error>()
+      getInsuranceContractsUseCase.addTerminatedInsurancesToResponse()
+      assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Loading>()
+      assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Success>()
```

Comment:
same question: why Loading state?

Reply:
If you move `      assertThat(awaitItem()).isInstanceOf<TerminatedContractsUiState.Loading>()
` to the line above the `addTerminatedInsurancesToResponse` the test still passes right?

From the code it looks like when the LaunchedEffect starts, it it's not success it will automatically set the state to loading. And here we start with error, so this does happen.
The key concept here is that LaunchedEffect will in fact run by itself on the first composition, and here we are coming to a fresh presenter just with an error last state. That's still a fresh composition, which is the same behavior people will experience using the app for real.

So practically, the loading state is emitted to the UI, it's just that in the test we have not looked at it yet, it's just there waiting since every emission in our presenter is put in the turbine, which is more or less a channel with all the emissions, one by one as they come.

So this is great actually! This test caches the fact that if we come back to this screen and we have had an error state before, what we try is immediately set it to loading and we try to load the state again.

Now the question is if we want to keep this behavior or not. I'd expect if someone comes *back* into an error screen it's okay to implicitly trigger the reload for them and put them to the loading state, so this test will in fact keep this behavior here, so if we ever break this we will know by the test failing.

One thing I would absolutely recommend is to move the assertion to above the `addTerminatedInsurancesToResponse` to properly indicate that it is not the one which triggers this behavior, but it happens by itself!


Once again, amazing with how well you've picked up writing all these tests already, and that such behaviors start surfacing by adding them, I am very impressed, great job!

Reply:
Thanks! And thanks for the explanation, I somehow missed the fact that it's not only the initial state that is Loading originally, but also the first dataIteration = 0 that triggers the same Loading state.

But it's interesting how it works with turbine and all